internal/datanode/channel_manager.go

```diff
@@ -0,0 +1,320 @@
+	defer m.inProgressWaiter.Done()
+
+	dataSyncService := getDataSyncService(m.dn, info, tickler)
+	if dataSyncService != nil {
+		dataSyncService.start()
+		m.runningFlowgraphs.Add(dataSyncService)
+		metrics.DataNodeNumFlowGraphs.WithLabelValues(fmt.Sprint(paramtable.GetNodeID())).Inc()
+	}
+
+	m.finishInprogressOperation(info.GetVchan().GetChannelName())
+}
+
+func (m *ChannelManager) finishInprogressOperation(channel string) {
+	m.guard.Lock()
+	defer m.guard.Unlock()
+	delete(m.watchInProgress, channel)
+}
+
+func getDataSyncService(node *DataNode, info *datapb.ChannelWatchInfo, tickler *tickler) *dataSyncService {
+	channelName := info.GetVchan().GetChannelName()
+	log := log.With(zap.String("channel", channelName))
+
+	channel := newChannel(
+		info.GetVchan().GetChannelName(),
+		info.GetVchan().GetCollectionID(),
+		info.GetSchema(),
+		node.rootCoord,
+		node.chunkManager,
+	)
+
+	dataSyncService, err := newDataSyncService(
+		node.ctx,
+		make(chan flushMsg, 100),
+		make(chan resendTTMsg, 100),
+		channel,
+		node.allocator,
+		node.dispClient,
+		node.factory,
+		info.GetVchan(),
+		node.clearSignal,
+		node.dataCoord,
+		node.segmentCache,
+		node.chunkManager,
+		node.compactionExecutor,
+		tickler,
+		node.GetSession().ServerID,
+		node.timeTickSender,
+	)
+	if err != nil {
+		log.Warn("fail to create new datasyncservice", zap.Error(err))
+		return nil
+	}
+
+	return dataSyncService
+}
+
+func (m *ChannelManager) ExecuteReleaseOperation(info *datapb.ChannelWatchInfo) {
+	m.guard.Lock()
+	defer m.guard.Unlock()
+	log.Info("Execute channel release operation", zap.String("channel", info.GetVchan().GetChannelName()))
+	channel := info.GetVchan().GetChannelName()
+	if m.runningFlowgraphs.exist(channel) {
+		m.runningFlowgraphs.release(channel)
+		return
+	}
+
+	// If request release for a not running flowgraph
+	if m.hasOp(channel, datapb.ChannelWatchState_ToWatch) {
+		// if inProgress, stop and delete from the manager
+		if t, ok := m.watchInProgress[channel]; ok {
+			t.close()
+			delete(m.watchInProgress, channel)
+			return
+		}
+
+		// if inQueue, remove from the queue
+		for i, info := range m.opInQueue {
+			if info.GetVchan().GetChannelName() == channel {
+				m.opInQueue = append(m.opInQueue[:i], m.opInQueue[i+1:]...)
+				return
+			}
+		}
+	}
+}
+
+func (m *ChannelManager) GetOperationProgress(channel string, state datapb.ChannelWatchState) int32 {
+	m.guard.RLock()
+	defer m.guard.RUnlock()
+	for _, info := range m.opInQueue {
+		if info.GetVchan().GetChannelName() == channel && info.GetState() == state {
+			return 0
+		}
+	}
+
+	if state == datapb.ChannelWatchState_ToWatch {
+		if m.runningFlowgraphs.exist(channel) {
+			return 100
+		}
+
+		if t, ok := m.watchInProgress[channel]; ok {
+			return t.progress.Load()
+		}
+
+		// indicating Failure
+		return -1
+	}
+
+	if state == datapb.ChannelWatchState_ToRelease {
+		if !m.runningFlowgraphs.exist(channel) && !m.hasOp(channel, datapb.ChannelWatchState_ToRelease) {
```

Comment:
Add a Close() and GracefulClose(), GracefulClose will be used only on rolling upgrade, when we Stop() a DataNode; normal Release a Channel will just close and return, supposed not to fail.

If Close() failed, it's a bug need to fix. However GracefulClose() might fail bug won't be considerred here because datanode.Stop() won't triggers DC's channel release operation.